pkg/applyconfigurations/gen.go

```diff
@@ -0,0 +1,383 @@
+/*
```

Comment:
Two somewhat abitrary datapoints for trying this out:

When using this with cluster api v1alpha4 (https://github.com/kubernetes-sigs/cluster-api/tree/0065d5c80a1be27e81250eae170915b88a8d53a7/api/v1alpha4) the generated applyconfig doesn't compile:
```
api/v1alpha4/ac/zz_generated.applyconfigurations.go:119:2: undefined: "k8s.io/client-go/applyconfigurations/meta/v1".ListMetaApplyConfiguration
api/v1alpha4/ac/zz_generated.applyconfigurations.go:173:2: undefined: "k8s.io/client-go/applyconfigurations/meta/v1".ListMetaApplyConfiguration
api/v1alpha4/ac/zz_generated.applyconfigurations.go:642:2: undefined: "k8s.io/client-go/applyconfigurations/meta/v1".ListMetaApplyConfiguration
api/v1alpha4/ac/zz_generated.applyconfigurations.go:903:2: undefined: "k8s.io/client-go/applyconfigurations/meta/v1".ListMetaApplyConfiguration
api/v1alpha4/ac/zz_generated.applyconfigurations.go:1042:2: undefined: "k8s.io/client-go/applyconfigurations/meta/v1".ListMetaApplyConfiguration
api/v1alpha4/ac/zz_generated.applyconfigurations.go:1123:2: undefined: "k8s.io/client-go/applyconfigurations/meta/v1".ListMetaApplyConfiguration
```

When using this with Prow v1 (https://github.com/kubernetes/test-infra/tree/febc5aef37857e5aef5140963e02f94c78ecabc1/prow/apis/prowjobs/v1) the generation fails altogether:
```
$ controller-gen apply paths=./prow/apis/prowjobs/v1
panic: interface conversion: types.Type is *types.Basic, not *types.Named

goroutine 1 [running]:
sigs.k8s.io/controller-tools/pkg/applyconfigurations.shouldBeApplyConfiguration(0xc0010f7d20, 0xc003fad3e0)
	/home/alvaro/git/golang/src/sigs.k8s.io/controller-tools/pkg/applyconfigurations/traverse.go:402 +0x372
sigs.k8s.io/controller-tools/pkg/applyconfigurations.(*ObjectGenCtx).generateEligibleTypes.func1(0xc003fad3e0)
	/home/alvaro/git/golang/src/sigs.k8s.io/controller-tools/pkg/applyconfigurations/gen.go:223 +0x3a
sigs.k8s.io/controller-tools/pkg/markers.EachType.func1(0xc000f1be00, 0xc0017c8d00, 0xc0017029c0)
	/home/alvaro/git/golang/src/sigs.k8s.io/controller-tools/pkg/markers/zip.go:179 +0x257
sigs.k8s.io/controller-tools/pkg/loader.(*typeVisitor).Visit(0xb, {0xcbd6f8, 0xc0017029c0})
	/home/alvaro/git/golang/src/sigs.k8s.io/controller-tools/pkg/loader/visit.go:64 +0xe9
go/ast.Walk({0xcb24a0, 0xc002b15f50}, {0xcbd6f8, 0xc0017029c0})
	/usr/local/go/src/go/ast/walk.go:50 +0x5f
go/ast.Walk({0xcb24a0, 0xc002b15f50}, {0xcbd360, 0xc0017c8d00})
	/usr/local/go/src/go/ast/walk.go:333 +0x1e49
go/ast.walkDeclList({0xcb24a0, 0xc002b15f50}, {0xc000b38e00, 0xb, 0xc00107cd40})
	/usr/local/go/src/go/ast/walk.go:36 +0x87
go/ast.Walk({0xcb24a0, 0xc002b15f50}, {0xcbd298, 0xc000f1be00})
	/usr/local/go/src/go/ast/walk.go:355 +0x15c5
sigs.k8s.io/controller-tools/pkg/loader.EachType(0xc0010f7d20, 0xc002b15f38)
	/home/alvaro/git/golang/src/sigs.k8s.io/controller-tools/pkg/loader/visit.go:38 +0x96
sigs.k8s.io/controller-tools/pkg/markers.EachType(0xc001131740, 0xc00113b888, 0xc002b15ec0)
	/home/alvaro/git/golang/src/sigs.k8s.io/controller-tools/pkg/markers/zip.go:155 +0x92
sigs.k8s.io/controller-tools/pkg/applyconfigurations.(*ObjectGenCtx).generateEligibleTypes(0xc0052cda48, 0xc0010f7d20)
	/home/alvaro/git/golang/src/sigs.k8s.io/controller-tools/pkg/applyconfigurations/gen.go:222 +0xb8
sigs.k8s.io/controller-tools/pkg/applyconfigurations.Generator.Generate({{0x0, 0xb25aa0}, {0x0, 0xb87e01}}, 0xc00009dae0)
	/home/alvaro/git/golang/src/sigs.k8s.io/controller-tools/pkg/applyconfigurations/gen.go:168 +0x53e
sigs.k8s.io/controller-tools/pkg/genall.(*Runtime).Run(0xc000cbd000)
	/home/alvaro/git/golang/src/sigs.k8s.io/controller-tools/pkg/genall/genall.go:207 +0x206
main.main.func1(0xc0001ffb80, {0xc00014bf20, 0x2, 0x2})
	/home/alvaro/git/golang/src/sigs.k8s.io/controller-tools/cmd/controller-gen/main.go:178 +0x6a
github.com/spf13/cobra.(*Command).execute(0xc0001ffb80, {0xc00011a160, 0x2, 0x2})
	/home/alvaro/git/golang/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:850 +0x60e
github.com/spf13/cobra.(*Command).ExecuteC(0xc0001ffb80)
	/home/alvaro/git/golang/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:958 +0x3ad
github.com/spf13/cobra.(*Command).Execute(...)
	/home/alvaro/git/golang/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:895
main.main()
	/home/alvaro/git/golang/src/sigs.k8s.io/controller-tools/cmd/controller-gen/main.go:202 +0x2c6
```

Reply:
clusterapiv1alpha4 should be fully working now. Prow types are a bit problematic because they haven't been updated in a while. controller-tools now depends on kubernetes 1.25+ and the types haven't been updated to use the latest version.